src/frequenz/client/electricity_trading/_client.py

```diff
@@ -159,6 +177,87 @@ def __init__(
 
         self._metadata = (("key", auth_key),) if auth_key else ()
 
+        # Timeout configuration
+        self._initial_timeout = initial_timeout
+        self._timeout = initial_timeout
+        self._timeout_increment = timeout_increment
+        self._max_timeout = max_timeout
+        self._max_timeout_retries = max_timeout_retries
+        self._reset_interval = reset_interval
+        self._last_success_time = datetime.now(timezone.utc)
+
+    def _increase_timeout(self) -> None:
+        """Increase the global timeout within limits."""
+        if self._timeout < self._max_timeout:
+            self._timeout = min(
+                self._max_timeout, self._timeout + self._timeout_increment
+            )
+            _logger.info("Increased timeout to %ds.", self._timeout)
+        else:
+            _logger.warning(
+                "Timeout is already at the maximum (%ds). No further increments.",
+                self._max_timeout,
+            )
+
+    def _reset_timeout_if_needed(self) -> None:
+        """Reset the timeout to the initial value if the reset interval has passed."""
+        now = datetime.now(timezone.utc)
+        if (now - self._last_success_time).total_seconds() >= self._reset_interval:
+            self._timeout = self._initial_timeout
+            _logger.info(
+                "Timeout reset to initial value %ds after %ds seconds without timeout errors.",
+                self._initial_timeout,
+                self._reset_interval,
+            )
+
+    async def grpc_call_with_timeout(
+        self,
+        call: Callable[..., Awaitable[Any]],
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
```

Comment:
I would suggest extracting the timeout logic to a separate `backoff` module, so that you can unit test it. You could just use that `backoff` module here.

Reply:
Done